### ISSUE TITLE
input and output  and number of blocks in this module

### ISSUE THREAD

**Issue by kianooshaminiankianooshaminian** [kuhnke-m/gr-mthpower#1]
i have downloaded the file     https://github.com/kuhnke-m/gr-mthpower      and installed that in gnu radio. but there is only two blocks in the module while mentioning 9 blocks in the current directory and also the input and output ports have the color white which is an unknown type for me.
i will be so thankful if anyone gives me some tips or advice on this problem.

**Comment by kuhnke-m** [comment]
Hi,
I just saw that the xml files were broken, I fixed them, hope the ports show the correct type now!

Generally: At the moment there are only two blocks in this module. They are:
- m_th_power_fc_cc (Frequency correction)
- m_th_power_v1_cc (Phase correction)
Maybe it helps you if I tell you what this module is about. 
Currently it can be used as a replacement for a Costas Loop for M-PSK Modulation formats.
It can remove residual carrier frequency offsets and phase offsets introduced during transmission.
The frequency correction ! limit ! is: Rs/(2M) (Where Rs=SymbolRate, M as in M-PSK)
Frequency offset should be well below that, otherwise the Bit Error Rate increases.

Here is what you need to do to get it working: 
Transmitter side:
- differentially encode your data
- map the data to QPSK Symbols
- shape them using a matched filter (i.e. RRC, alpha=0.8)
- feed into IQ Mod
Receiver side:
- IQ Demod
- match filter (i.e. RRC)
- process samples using DC Block, AGC, (maybe Channel Equalizer), etc
- feed into Symbol Sync block (e.g. Gardner Algorithmus, decimate to 1 sample/symbol)
- feed into m-th-power frequency correction (m_th_power_fc_cc, you can use large N)
- then feed into m-th-power phase correction (m_th_power_v1_cc)
- After that, you can map symbols to data (Constellation Decode)
- Then differentially decode

The m_th_power_fc_cc block has one input for complex samples, one float output (orange) for debug output ( estimated frequency), one output for the corrected samples. 

The m_th_power_v1_cc block has one input for a vector of complex samples and one output a vector.
If you use a Blocklen N different from 1, then you need to convert samples to a vector at the input and output port to be able to use it.

Good luck

**Comment by kianooshaminiankianooshaminian** [comment]
Hi dear friend 
thank you so much for helping me on this problem. that was so much useful.
thanks again. 
bests

**Comment by kianooshaminiankianooshaminian** [comment]
Hi
actually the above code works only when i set the vector size of the block mthpower_m_th_power_v1_cc.xml to 4. indeed the flow graph does not work for any vector size except  4. and the error is such as below.
it says that the block mthpower_m_th_power_v1_cc.xml only works with 32 bits.
and also the result of the flow graph with vector size of 4 is only acceptable for Qpsk not any other Mpsk
i will be thank full if you answer or give me some solution or correct the code.

**Comment by kuhnke-m** [comment]
Hi, 

I will look into the issue with different vector sizes not working.
However. The vector size parameter N is not related to M (as in M-PSK), it should (not tested, although very probable) work for any M-PSK. What it does is, it gathers N samples (in this case symbols, because there must be one symbol/sample) and applies the algorithm to all of them. 
The hypothesis was, that this can speed up processing of symbols/second. Larger vector sizes can decrease phase noise impact on BER, however it is a very moderate improvement.
If performance is not an issue the vector size can be set to one or left at 4. It doesn't matter and should work for any M-PKS.

I will test the vector issues, and maybe test it for 8-PSK too just to be sure, will report back when I am done :)

Regards

**Comment by kuhnke-m** [comment]
Hi,
I fixed the vector length error message (see latest commit). 
I also checked the operation of the phase recovery in isolation and in combination with the frequency correction for 8-PSK using a simple flow graph and using a testbench function. 
It works, even for different block sizes. It matches values predicted by formulas and correctly recovers the data from the symbols.
So these modules are ~99.9% working for QPSK as well as 8-PSK and most likely (>99% confidence) for every other PSK.

I commited the simple test flow graph for your reference "examples/test.grc"

Regards
